beanie/odm/utils/encoder.py

```diff
@@ -1,119 +1,140 @@
-                                obj_dict[k] = o
-                    else:
-                        obj_dict[k] = o
-            return self.encode(obj_dict, custom_encoder=encoders, to_db=to_db)
-        if isinstance(obj, BaseModel):
-            encoders = {}
-            if custom_encoder:
-                encoders.update(custom_encoder)
-            obj_dict = {}
-            for k, o in obj._iter(to_dict=False, by_alias=by_alias):
-                if k not in exclude:  # TODO get exclude from the class
+                        else:
+                            obj_dict[k] = o
+                else:
                     obj_dict[k] = o
+                obj_dict[k] = encoder.encode(obj_dict[k])
+        return obj_dict
 
-            return self.encode(obj_dict, custom_encoder=encoders, to_db=to_db)
-        if isinstance(obj, dict):
-            encoded_dict = {}
-            for key, value in obj.items():
-                encoded_value = self.encode(
-                    value,
-                    by_alias=by_alias,
-                    custom_encoder=custom_encoder,
-                    to_db=to_db,
-                )
-                encoded_dict[key] = encoded_value
-            return encoded_dict
-        if isinstance(obj, (list, set, frozenset, GeneratorType, tuple)):
-            return [
-                self.encode(
-                    item,
-                    exclude=exclude,
-                    by_alias=by_alias,
-                    custom_encoder=custom_encoder,
-                    to_db=to_db,
-                )
-                for item in obj
-            ]
-        if custom_encoder:
-            if type(obj) in custom_encoder:
-                return custom_encoder[type(obj)](obj)
-            for encoder_type, encoder in custom_encoder.items():
+    def encode_base_model(self, obj):
+        obj_dict = {}
+        for k, o in obj._iter(to_dict=False, by_alias=self.by_alias):
+            if k not in self.exclude:
+                obj_dict[k] = self._encode(o)
+
+        return obj_dict
+
+    def encode_dict(self, obj):
+        for key, value in obj.items():
+            obj[key] = self._encode(value)
+        return obj
+
+    def encode_iterable(self, obj):
+        return [self._encode(item) for item in obj]
+
+    def _encode(
+        self,
+        obj,
+    ) -> Any:
+        from beanie.odm.documents import Document
+
+        if self.custom_encoders:
+            if type(obj) in self.custom_encoders:
+                return self.custom_encoders[type(obj)](obj)
+            for encoder_type, encoder in self.custom_encoders.items():
                 if isinstance(obj, encoder_type):
                     return encoder(obj)
         if type(obj) in ENCODERS_BY_TYPE:
             return ENCODERS_BY_TYPE[type(obj)](obj)
-        for encoder, classes_tuple in self.encoders_by_class_tuples.items():
-            if isinstance(obj, classes_tuple):
+        for cls, encoder in ENCODERS_BY_TYPE.items():

```

Comment:
Hi,
Can u explain why this loop need?

I was able to resolve [this error](https://github.com/roman-right/beanie/issues/209)
But this solution returns unexpected value when I write a model that has datetime.datetime and datetime.date

You can reproduce by this code

```python
class MyDoc(beanie.Document):
    x: datetime.datetime
    y: datetime.date

    class Settings:
        bson_encoders = {datetime.date: lambda x: datetime.datetime.combine(x, datetime.time.min)}

doc = MyDoc(x=datetime.datetime.now(), y=datetime.date.today())
await doc.save()
doc = MyDoc.fnid_one(MyDoc.id == doc.id)
print(doc)
# id=ObjectId('627e8b3e228faaef81209a64') revision_id=UUID('a203d6a0-1180-4f71-b643-f48b0e6696b1') x=datetime.datetime(2022, 5, 14, 0, 0) y=datetime.date(2022, 5, 14)
```
Cause: datetime.datetime subclass of datetime.date
```python
isinstance(datetime.datetime.now(), datetime.date)
# True
```

I can solve using work-around (`add datetime.datetime encoder`) but I cant understand why need additional loop for finding encoder